### ISSUE TITLE
`idris-stay-in-current-window-on-compiler-error` has no effect

### ISSUE THREAD

**Issue by ExpHP** [syl20bnr/spacemacs#8354]
#### Description :octocat:
The `idris-stay-in-current-window-on-compiler-error` configuration option of `idris-mode` does not have any effect in spacemacs, despite working fine in standard emacs with idris-mode.

#### Reproduction guide :beetle:
- Install the `idris` layer
- Have the following in dotspacemacs/user-config

```elisp
  (with-eval-after-load 'idris-mode
    (setq idris-stay-in-current-window-on-compiler-error t))
```

- Create a file `test.idr` with a type error:

```idris
a : Nat
a = 0.2
```

- Start Emacs
- Open `test.idr`
- `,r` to send the file to the interpreter.
- Windows `*idris-repl*` and `*idris-notes*` will appear _and_...

*Observed behaviour:* :eyes: :broken_heart:
...the notes window steals input focus!

*Expected behaviour:* :heart: :smile:
...window `#1` retains input focus.

#### System Info :computer:
- OS: gnu/linux
- Emacs: 24.5.1
- Spacemacs: 0.200.7
- Spacemacs branch: develop (rev. 57e653a)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(helm emacs-lisp idris org)
```

#### Additional notes: :clipboard:

The code that handles this config option in idris-mode is [here](https://github.com/idris-hackers/idris-mode/blob/fa743b77b8760a/idris-commands.el#L238-L243). If I change the following line:

```elisp
(display-buffer (idris-buffer-name :notes))
```

to

```elisp
(save-selected-window (display-buffer (idris-buffer-name :notes)))
```

then the desired behavior is recovered.

So why, then, am I bugging the spacemacs crew? _Because I tested it and found that **the option works just fine without this modification outside spacemacs**..._ and I'm not quite sure what to make of that!

That is, if I install `idris-mode-20161102.707` (the same version used in spacemacs) onto a non-spacemacs, blank-slate emacs, and try loading (`C-c C-l`) my test file, then the error window that is created either does or does not steal focus, honoring the value of `idris-stay-in-current-window-on-compiler-error`.

**Comment by ExpHP** [comment]
**Update:** I believe [the `popwin` configuration in the `+lang/idris` layer](https://github.com/syl20bnr/spacemacs/blob/3290c8a94f5/layers/%2Blang/idris/packages.el#L119-L127) is to blame.  Thankfully this can be easily overriden from the user config:

```elisp
(defun dotspacemacs/user-config ()
  ; ...
  (with-eval-after-load 'idris-mode
    (setq idris-stay-in-current-window-on-compiler-error t)
    (dolist (x '("*idris-notes*" "*idris-holes*" "*idris-info*"))
      (plist-put (cdr (assoc x popwin:special-display-config)) :noselect t))
  )
)
```

which not only recovers the desired behavior, but makes the Holes window more polite as well. (which means I can finally stop disabling it!)

I'm satisfied with these circumstances; if others don't feel action is necessary, this can be closed.

**Comment by ExpHP** [comment]
(closing since it's no big deal to me and there is already a massive backlog of issues)

**Comment by ElessarWebb** [comment]
This is super useful! Without this Idris-mode is incredibly annoying to use as one has to navigate back to the code after loading it. Shouldn't this just be fixed?

**Comment by ExpHP** [comment]
I'm glad to see that somebody else agrees!  I'll reopen the issue.

I'm satisfied enough with the workaround that I don't feel any driving need to submit a PR; but it would be great to see this fixed if somebody is up to it!

**Comment by ElessarWebb** [comment]
I have zero emacs knowledge, so I won't be much use in fixing this. But I
agree that this should remain open until a proper fix exists.

**Comment by bmag** [comment]
We can probably change the `:noselect` value in our configuration. I hope to fix this myself soon, but a PR is also welcome. Need to make sure that `:noselect t` doesn't prevent the popup from being selected when `idris-stay-in-current-window-on-compiler-error` is `nil`.

**Comment by ExpHP** [comment]
I did some testing on standard emacs to see how idris-mode behaves, versus spacemacs with `idris/packages.el` modified to use `:noselect t`.  I must stress that I am not a big idris user (after all, this issue was just something I ran into when trying the language out!); just somebody poking at various features to see if things look alright.

NOTE: this comparison does not include the current behavior of spacemacs

---

<details>
<summary>What I tested</summary>

```idris
-- ok.idr
a : Nat
a = 2

-- hole.idr
a : Nat
a = ?b

-- error.idr
a : Nat
a = 2.0
```

**Key:**
:white_check_mark: = spacemacs works same (or "uncontestably" better) than plain idris-mode
:x: = spacemacs is clearly broken
:question: = spacemacs is different but not clearly broken

**Tests:**

**Type error (default config)** :white_check_mark:
- idris-mode test: open `error.idr`. `C-c C-l`
- spacemacs test: open `error.idr`. `, r`
- **idris-mode:** idris-notes appears and steals focus. Can interact, or close with q.
- **noselect t:** idris-notes appears and steals focus. Can interact, or close with q.

**Type error (idris-stay-in-current-window-on-compiler-error t)** :white_check_mark:
- same test as before, but with the config variable set beforehand.
- **idris-mode:** idris-notes appears, does not steal focus. Cursor moves to line with error.
- **noselect t:** idris-notes appears, does not steal focus. Cursor moves to line with error. 

**idris-type-at-point** :white_check_mark:
- idris-mode test: open `fine.idr`.  `C-c C-l` `C-c C-t`
- spacemacs test: open `fine.idr`. `, r` `, t`
- **idris-mode:** idris-info appears, steals focus. Close with `q`.
  - (note: I found that, in some conditions, closing idris-info put focus in idris-repl instead of the code; seemed unpredictable)
- **noselect t:** idris-info appears, steals focus. Close with `q`.
  - (note: this also had issues with getting sent to the repl, but here it consistently occurs only on the first test.  If I click back into the code, then doing `, t q` a second time will return me to the code)

**Holes window** :white_check_mark:
- idris-mode test: open `hole.idr`. `C-c C-l`
- spacemacs test: open `hole.idr`. `, r`
- **idris-mode:** idris-holes appears but does not steal focus.
- **noselect t:** idris-holes appears but does not steal focus.

**idris-name-calls-who-at-point** :x:
- I more or less pulled this one out of a hat
- idris-mode test: open `ok.idr`. `C-c C-l ` `M-x idris-name-calls-who-at-point`
- spacemacs test: open `ok.idr`. `, r ` `M-x idris-name-calls-who-at-point`
- **noselect t:** idris-tree-viewer appears, steals focus. Close with q.
  - (this usually replaces it with the repl, which takes focus. blech)
- **noselect t:** idris-tree-viewer appears, steals focus. Statusbar says "q to close" **but it does not close.**
  - (After doing `M-x evil-motion-state`, then `q` replaces the buffer with `*scratch*`. Still blech.)

</details>

---

In summary it looks like:

* `noselect t` **does not interfere** with events that *should* steal focus. (hooray!)
* There may be other buffers that should probably receive some or all of the same treatment that `*idris-notes*`, `*idris-holes*`, and `*idris-info*` currently receive. (notably `*idris-tree-viewer*`)

**Comment by ExpHP** [comment]
Ah rats.  After more testing it's not quite the same:

The current behavior of spacemacs is to refocus the error window on every reload that has errors.  Plain idris-mode under default settings behaves in this same manner.

With `:noselect nil` under default settings, however, it only focuses the error window when it is newly created.  If I leave the `notes` window open, do more editing and `, r` again, it leaves the code selected.  This is inconsistent and confusing behavior.

---

I'm really not sure what a proper fix is, but it looks like the popwin config for `*idris-notes*` will need to somehow change in response to `idris-stay-in-current-yadda-yada-yadda`

**Comment by bmag** [comment]
> With `:noselect nil` under default settings

Did you mean `:noselect t`? With `:noselect t`, Spacemacs doesn't force the focus, so this is entirely up to idris-mode to decide (I'm saying this based on reading the code for Spacemacs and idris-mode).

Below is what I think is a proper fix, but I don't know where the "inconsistent and confusing behavior" comes from. My best estimate is that idris-mode doesn't try to display the notes buffer in this case.
```diff
diff --git a/layers/+lang/idris/packages.el b/layers/+lang/idris/packages.el
index fe18bb8..cf89c63 100644
--- a/layers/+lang/idris/packages.el
+++ b/layers/+lang/idris/packages.el
@@ -121,9 +121,9 @@
 (defun idris/pre-init-popwin ()
   (spacemacs|use-package-add-hook popwin
     :post-config
-    (push '("*idris-notes*" :dedicated t :position bottom :stick t :noselect nil :height 0.4)
+    (push '("*idris-notes*" :dedicated t :position bottom :stick t :noselect t :height 0.4)
           popwin:special-display-config)
-    (push '("*idris-holes*" :dedicated t :position bottom :stick t :noselect nil :height 0.4)
+    (push '("*idris-holes*" :dedicated t :position bottom :stick t :noselect t :height 0.4)
           popwin:special-display-config)
-    (push '("*idris-info*" :dedicated t :position bottom :stick t :noselect nil :height 0.4)
+    (push '("*idris-info*" :dedicated t :position bottom :stick t :noselect t :height 0.4)
           popwin:special-display-config)))
```

**Comment by ExpHP** [comment]
Oops, you are right, I meant :noselect t; and those are precisely the changes I tested.  The trouble is that, in testing, I found that in idris-mode and current spacemacs, the errors window really is *forced* into focus, but in my topic branch, it only takes focus *as a suggestion.*

My naive guess is that it has something to do with the other flags like maybe `dedicated`. I would have to test more.
